File: src/main/java/com/letgo/book/infrastructure/configuration/HibernateConfiguration.java
package com.letgo.book.infrastructure.configuration;

import org.springframework.boot.jdbc.DatabaseDriver;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.io.FileSystemResource;
import org.springframework.jdbc.datasource.DriverManagerDataSource;
import org.springframework.orm.hibernate5.HibernateTransactionManager;
import org.springframework.orm.hibernate5.LocalSessionFactoryBean;
import org.springframework.transaction.PlatformTransactionManager;
import org.springframework.transaction.annotation.EnableTransactionManagement;

import javax.sql.DataSource;
import java.util.Properties;

@Configuration
@EnableTransactionManagement
public class HibernateConfiguration {
    @Bean
    public LocalSessionFactoryBean sessionFactory() {
        LocalSessionFactoryBean sessionFactory = new LocalSessionFactoryBean();
        sessionFactory.setDataSource(dataSource());
        sessionFactory.setMappingDirectoryLocations(
                new FileSystemResource("./src/main/java/com/letgo/book/infrastructure/persistence/mapping")
        );
        sessionFactory.setHibernateProperties(hibernateProperties());
        return sessionFactory;
    }

    @Bean
    public DataSource dataSource() {
        DriverManagerDataSource dataSource = new DriverManagerDataSource();
        dataSource.setDriverClassName(DatabaseDriver.MYSQL.getDriverClassName());
        dataSource.setUrl("jdbc:mysql://java-app-demo-mysql:3306/demo");
        dataSource.setUsername("root");
        dataSource.setPassword("");
        return dataSource;
    }

    @Bean
    public PlatformTransactionManager hibernateTransactionManager() {
        HibernateTransactionManager transactionManager = new HibernateTransactionManager();
        transactionManager.setSessionFactory(sessionFactory().getObject());
        return transactionManager;
    }

    private Properties hibernateProperties() {
        Properties properties = new Properties(2);
        properties.put("hibernate.dialect", "org.hibernate.dialect.MySQL8Dialect");
        properties.put("hibernate.hbm2ddl.auto", "none");
        return properties;
    }
}

File: src/main/java/com/letgo/book/infrastructure/configuration/HibernateConfiguration.kt
package com.letgo.book.infrastructure.configuration

import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.boot.jdbc.DatabaseDriver
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.core.io.FileSystemResource
import org.springframework.jdbc.datasource.DriverManagerDataSource
import org.springframework.orm.hibernate5.HibernateTransactionManager
import org.springframework.orm.hibernate5.LocalSessionFactoryBean
import org.springframework.transaction.PlatformTransactionManager
import org.springframework.transaction.annotation.EnableTransactionManagement
import java.util.Properties
import javax.sql.DataSource

@Configuration
@EnableTransactionManagement
@EnableConfigurationProperties(DatabaseParameters::class)
open class HibernateConfiguration(
    private val dbParameters: DatabaseParameters,
) {
    @Bean
    open fun sessionFactory(): LocalSessionFactoryBean = LocalSessionFactoryBean().apply {
        setDataSource(dataSource())
        setMappingDirectoryLocations(
            FileSystemResource("./src/main/java/com/letgo/book/infrastructure/persistence/mapping")
        )
        hibernateProperties = hibernateProperties()
    }

    @Bean
    open fun dataSource(): DataSource = DriverManagerDataSource().apply {
        setDriverClassName(DatabaseDriver.MYSQL.driverClassName)
        url = "jdbc:mysql://${dbParameters.host}:${dbParameters.port}/${dbParameters.name}"
        username = dbParameters.user
        password = dbParameters.password
    }

    @Bean
    open fun hibernateTransactionManager(): PlatformTransactionManager = HibernateTransactionManager().apply {
        sessionFactory = sessionFactory().getObject()
    }

    private fun hibernateProperties(): Properties = Properties(2).apply {
        setProperty("hibernate.dialect", "org.hibernate.dialect.MySQL8Dialect")
        setProperty("hibernate.hbm2ddl.auto", "none")
    }
}
